vowpalwabbit/parser.h

```diff
@@ -35,6 +37,46 @@ struct example_initializer
   example* operator()(example* ex);
 };
 
+struct IO_State {
+
+      std::queue<IO_Item> *io_lines = nullptr;
+      std::atomic<bool> done_with_io;
+
+      IO_State(){
+        io_lines = new std::queue<IO_Item>;
+         done_with_io.store(false);
+      }
+
+      IO_State(std::queue<IO_Item> *new_input_lines){
+          //input_lines_copy now points to new_input_lines
+          io_lines = new_input_lines;
+          done_with_io.store(false);
+      }
+
+      IO_State operator=(const IO_State &toCopy){
+          io_lines = toCopy.io_lines;
+          done_with_io.store(toCopy.done_with_io);
+          return *this;
+      }
+
+      IO_State(const IO_State &toCopy){
+          io_lines = toCopy.io_lines;
+          done_with_io.store(toCopy.done_with_io);
+      }
+
+      inline void set_done_io(bool done_io){
+        done_with_io.store(done_io);
+      }
+
+      inline bool get_done_io(){
+        return done_with_io;
+      }
+
+
+      ~IO_State() {}
```

Comment:
Currently this is leaking the `io_lines` object. Using a `std::unique_ptr` is a quick and easy way to resolve this.